### PR TITLE
Cluster related metrics, user-agent

### PR DESCRIPTION
**Pull request by hcaumeil**
- Renaming gauge metrics that contains "count" in their name for consistency
- Make back-end HTTP status metrics,  5.x.x sozu HTTP status metrics and access_logs metric cluster-related (see #937 ).
- Add user agent in access logs (see #937 ).
- Add time metrics related to path rules / domain rules matching (see #804  ).

### CODE REVIEW

File: doc/debugging_strategies.md [hunk 1]
```diff
@@ -131,16 +131,16 @@ unresponsive backend server).
...
 * `sozu.http.400.errors`: cannot parse hostname
 * `sozu.http.404.errors`: unknown hostname and/or path
 * `sozu.http.413.errors`: request too large
 * `sozu.http.503.errors`: could not connect to backend server, or no backend server available for the corresponding cluster
 
 Going further, backend connections issues are tracked by the following metrics:
 
-* `sozu.backend.connections.errors`: could not connect to a backend server
-* `sozu.backend.down`: the retry policy triggered and marked the backend server as down
+* `sozu.connections.errors`: could not connect to a backend server
```
Comment (FlorentinDUBOIS) [inline doc/debugging_strategies.md:142]: for this one, we should keep the backend in name as it is related.

**Review by FlorentinDUBOIS — CHANGES_REQUESTED** [review]

File: lib/src/http.rs [hunk 2]
```diff
@@ -447,9 +448,26 @@ impl L7ListenerHandler for HttpListener {
...
+                    "frontend_matching_time",
+                    &c,
+                    (now - start).whole_milliseconds()
+                );
+            }
+        } else {
+            incr!("http.failed_backend_matching")
+        }
+
+        res
```
Comment (Keksoj) [inline lib/src/http.rs:470]: please rename `res` to `result` (if present at all), `r` to `route`, and `c` to `cluster`.

File: lib/src/http.rs [hunk 3]
```diff
@@ -447,9 +448,26 @@ impl L7ListenerHandler for HttpListener {
         */
         let host = unsafe { from_utf8_unchecked(hostname) };
 
-        self.fronts
+        let res = self
+            .fronts
             .lookup(host.as_bytes(), uri.as_bytes(), method)
-            .ok_or(FrontendFromRequestError::NoClusterFound)
+            .ok_or(FrontendFromRequestError::NoClusterFound);
```
Comment (Keksoj) [inline lib/src/http.rs:454]: I suggest doing:

```
        let route = self
            .fronts
            .lookup(host.as_bytes(), uri.as_bytes(), method)
            .ok_or_else(|| {
                incr!("http.failed_backend_matching");
                FrontendFromRequestError::NoClusterFound
            })?;
```
and drop the `if let Ok(route)` after.

File: lib/src/https.rs [hunk 4]
```diff
@@ -584,9 +585,26 @@ impl L7ListenerHandler for HttpsListener {
...
+                    "frontend_matching_time",
+                    &c,
+                    (now - start).whole_milliseconds()
+                );
+            }
+        } else {
+            incr!("http.failed_backend_matching")
+        }
+
+        res
```
Comment (Keksoj) [inline lib/src/https.rs:607]: same remarks as for `http.rs`

File: lib/src/metrics/mod.rs [hunk 5]
```diff
@@ -321,11 +321,13 @@ macro_rules! count (
 macro_rules! incr (
   ($key:expr) => (count!($key, 1));
   ($key:expr, $cluster_id:expr, $backend_id:expr) => {
-    use $crate::metrics::Subscriber;
+    {
```
Comment (Keksoj) [inline lib/src/metrics/mod.rs:324]: nice

File: lib/src/protocol/kawa_h1/mod.rs [hunk 6]
```diff
@@ -788,15 +789,49 @@ impl<Front: SocketHandler, L: ListenerHandler + L7ListenerHandler> Http<Front, L
...
+                        self.backend_id.as_deref()
+                    )
+                }
+                DefaultAnswerStatus::Answer504 => {
+                    incr!(
+                        "http.504.errors",
+                        self.cluster_id.as_deref(),
+                        self.backend_id.as_deref()
+                    )
+                }
```
Comment (Keksoj) [inline lib/src/protocol/kawa_h1/mod.rs:834]: are we obligated to have all these additionnal curly brackets? They seem a bit overkill.
Just a cosmetic remark.

File: lib/src/router/mod.rs [hunk 7]
```diff
@@ -419,7 +420,13 @@ impl DomainRule {
                     && !&hostname[..len_without_suffix].contains(&b'.')
             }
             DomainRule::Exact(s) => s.as_bytes() == hostname,
-            DomainRule::Regex(r) => r.is_match(hostname),
+            DomainRule::Regex(r) => {
+                let start = Instant::now();
+                let res = r.is_match(hostname);
+                let now = Instant::now();
+                time!("regex_matching_time", (now - start).whole_milliseconds());
+                res
```
Comment (Keksoj) [inline lib/src/router/mod.rs:428]: `res` is mileading. Please rename to `is_a_match` or anything that conveys the idea of a boolean.

File: lib/src/router/mod.rs [hunk 8]
```diff
@@ -494,7 +501,12 @@ impl PathRule {
                 }
             }
             PathRule::Regex(r) => {
-                if r.is_match(path) {
+                let start = Instant::now();
+                let res = r.is_match(path);
+                let now = Instant::now();
+                time!("regex_matching_time", (now - start).whole_milliseconds());
+
+                if res {
```
Comment (Keksoj) [inline lib/src/router/mod.rs:509]: same remark, please rename `res` to a more explicit name.

**Review by Keksoj — CHANGES_REQUESTED** [review]
This seems very okay. I request changes for variable naming and code presentation, for a better legibility.

Thank you for contributing!

**Review by Keksoj — APPROVED** [review]
everything looks great to me now. Thanks!

File: doc/debugging_strategies.md [hunk 9]
```diff
@@ -131,16 +131,16 @@ unresponsive backend server).
...
+* `sozu.http.frontend_parse_errors`: sozu received some invalid traffic
 * `sozu.http.400.errors`: cannot parse hostname
 * `sozu.http.404.errors`: unknown hostname and/or path
 * `sozu.http.413.errors`: request too large
 * `sozu.http.503.errors`: could not connect to backend server, or no backend server available for the corresponding cluster
 
 Going further, backend connections issues are tracked by the following metrics:
 
-* `sozu.backend.connections.errors`: could not connect to a backend server
-* `sozu.backend.down`: the retry policy triggered and marked the backend server as down
```
Comment (FlorentinDUBOIS) [inline doc/debugging_strategies.md:143]: Could you keep the `.backend` here also?

**Review by FlorentinDUBOIS — APPROVED** [review]

**Pull request merged**